### DISCUSSION TITLE
Arbitrary Breakpoints

### DISCUSSION THREAD

**Discussion by browner12** [tailwindlabs/tailwindcss#8261] [Ideas]
Currently we can use arbitrary values for those rare situations we need to break out of our design system

```html
<div class="p-[11px] sm:p-[15px]"></div>
```

I think it'd also be helpful to occasionally use custom breakpoints as well.

```html
<div class="p-[11px] [400]:p-[13px] sm:p-[15px]"></div>
```

I have one very specific use case for this. When our main menu is below about 400px, we just have too much stuff in it and we have wrapping issues, but I don't want those things to get the adjusted style between 400px and SM.

I am a *little* hesitant to propose this idea, because if abused you could definitely lose out on some benefits of Tailwind, but I figured I'd throw it out to the group.

**Comment by jenstornell** [comment]
I like this idea. It could speed up development. At least until container queries arrive in CSS/Tailwind.

The downside would be that we could end up with very many breakpoints, a "breakpoint hell". But that's the downside to arbitrary values in general.

**Comment by theAdhocracy** [comment]
To add another use-case: we have two images, one with a portrait layout and the other with a horizontal one. Depending on the layout, we want to switch between the two. This works, but on our narrowest layout (single column) there are still plenty of devices that could fit the wider image, which would reduce the overall page height (and, frankly, look better).

There's no point adding a custom breakpoint for this single component, but it would be nice to have that slight flexibility in the editorial control.

**Comment by adamwathan** [comment]
You can do this now using arbitrary variants 👍 

https://play.tailwindcss.com/XaNfzyHXsr

```html
<div class="h-64 bg-red-500 [@media(min-width:711px)]:bg-green-500"></div>
```

Syntax is a bit verbose so maybe something we could simplify more in the future but it's at least possible when you need that escape hatch 👍

**Reply by benatshippabo** [reply]
@adamwathan this is awesome! My team recently did this, but spent a while debugging why it wasn't working until we finally figured out it was because we had a space in the arbitrary value. Any chance we could get some sort of linting to enhance the DX? https://github.com/tailwindlabs/tailwindcss/discussions/8863

**Reply by ji-horak** [reply]
That would be perfect. Before Tailwind, I was using my own mixin in SCSS called _bpmin_ (or bpmax) with input value, and it could be used exactly for those tweaking purposes.

**Reply by totti-rdz** [reply]
This worked perfectly fine! Thanks.
But the tailwind docs mention another approach for [arbitrary breakpoints](https://tailwindcss.com/docs/responsive-design#arbitrary-values), but this did not work for me. 
Can someone confirm, that this approach is not working for them, too? Maybe I made a mistake and it is actually working.

**Reply by chamberlainpi** [reply]
@totti-rdz can confirm, I cannot get **arbitrary breakpoints** to work either. Does something need to be enabled first in the tailwind config file? 🤔

**Reply by lymau** [reply]
@totti-rdz @chamberlainpi Yes, it's not working. I can confirm that. For me, I use the code that @adamwathan wrote above.

**Reply by RobinMalfait** [reply]
Both the snippet defined by @adamwathan and the snippet in the docs should work as expected: https://play.tailwindcss.com/xNsl2tXfp8

If this is not the case for you make sure that:
- You are on the latest version of Tailwind CSS
- You don't have any spaces in any of the classes

If something is still not working, could any of you create a minimal reproduction repo or CodeSandbox or Tailwind Play (https://play.tailwindcss.com) and open a new issue?

**Reply by totti-rdz** [reply]
Thanks @RobinMalfait for the tipps. This got quite confusing now. I updated Tailwind manually in my project and it didn't worked.
I created an example on Tailwind Play and examples with create-react-app+create-next-app, but there it worked.

In my project where it didn't worked I ran

`npm install -D tailwindcss postcss autoprefixer`

And it works now. I tried to revert the changes (and deleted node_modules) to update postcss and autoprefixer seperately to find the "culprit" package. Everytime deleting node_modules between the steps. But after reverting it still worked to my surprise.

I tested a little bit more and I found out, that only some classes do not work with arbitrary values according to the docs e.g. the flex-direction `flex-row`. I updated solely autoprefixer and now it works.

So I guess the solution is to update the autoprefixer package. And optionally the tailwind and postcss packages, so there is no compatibility issue with the new autoprefixer package.

@chamberlainpi and @lymau can you confirm that updating autoprefixer, maybe along with tailwindcss and postcss fixes this issue for you?

**Reply by MarvinMiles** [reply]
Updating autoprefixer/postcss did not worked for me.

At some point I got console warning:
> The `min-*` and `max-*` variants are not supported with a `screens` configuration containing objects.

So removing extended screen configs fixed the issue.

**Reply by aamirbinabd** [reply]
> So I guess the solution is to update the autoprefixer package. And optionally the tailwind and postcss packages, so there is no compatibility issue with the new autoprefixer package.

Had a similar issue. Updated both `postcss` and `autoprefixer` to latest version but had no effect. 

> So removing extended screen configs fixed the issue.

Also removed extended screen configs, checked for any issues in the file. Still getting this console warning:  

>warn - The `min-*` and `max-*` variants are not supported with a `screens` configuration containing objects.

Relying on `[@media(min-width:711px)]:bg-green-500` route for now.

**One possibly important note:**
I have configured to use Max-width breakpoints as explained in the docs [Max-width breakpoints](https://tailwindcss.com/docs/screens#max-width-breakpoints)

**Reply by wongjn** [reply]
> I have configured to use Max-width breakpoints

Yep, this stops `min-*`/`max-*` from working. See #9558 for details.

**Reply by shrekuu** [reply]
Could you please add this to this page, at the bottom? https://tailwindcss.com/docs/responsive-design

Maybe also add this warning I found in the terminal output: `warn - The `min-*` and `max-*` variants are not supported with a `screens` configuration containing objects.`

**Comment by sonsoixam** [comment]
i get same problem , any solution ?

**Reply by wongjn** [reply]
For Tailwind ≥ v3.0.0 and < v3.2.0, you can use an [arbitrary media query variant](https://github.com/tailwindlabs/tailwindcss/discussions/8261#discussioncomment-3251527).

For Tailwind ≥ v3.2.0, you can use an [arbitrary `min-*` or `max-*` variant](https://tailwindcss.com/docs/responsive-design#arbitrary-values).

**Reply by TristanBrotherton** [reply]
Same issue here - using TW `3.3.6` only the syntax `[@media(min-width:*)]:` will work not `min-*`

**Reply by wongjn** [reply]
Consider checking you only have "simple" `screen` values. As detailed in #9558:

> ## Your screens, if customized, MUST be `min-width` only and string values
> To ensure proper sorting we've made the decision to restrict the use of these features to screens that are "simple", meaning that they are `min-width`-based, and consist of only string values that use the same units.

**Reply by TristanBrotherton** [reply]
I can confirm in my `tailwind.config.cjs` there is no values set for screens.

**Reply by wongjn** [reply]
Consider checking:
- Tailwind has compiled since you added the class.
- The `content` file globs sufficiently cover the file that the class appears in.
- [String concatenation/interpolation is not used to build the class name](https://tailwindcss.com/docs/content-configuration#dynamic-class-names). 
- You are using the correct syntax for the class name.

**Reply by TristanBrotherton** [reply]
Thank you for the pointers a restart of vite seemed to do the trick.

**Comment by mauriciabad** [comment]
Now you can use the new range syntax:

```tsx
<div class="h-64 bg-red-500 [@media(width>=400px)]:bg-green-500"></div>
```

**Comment by logemann** [comment]
came here because the docs mention to use:

`min-[320px]:text-center`

Is this still valid? Doesnt work for me with latest TWCSS. And then, which one should i use?

**Reply by rohitmathur-7** [reply]
Yes, I also tried using arbitrary breakpoints like this: `min-[320px]:text-center`, but this does not work.

`<div class="h-64 bg-red-500 [@media(width>=400px)]:bg-green-500"></div>`
Using arbitrary breakpoints like this is working.

**Comment by jkinley** [comment]
I have confirmed that none of these work for me.

**Reply by andynghiaa** [reply]
className="hidden [@media(min-width:1400px)]:flex"
Please use this approach, it works well

**Comment by nemanjam** [comment]
```ts
      class: `border-2 max-[${TW_SCREENS.XS}px]:border-red-500 max-[${TW_SCREENS.MD}px]:border-green-500  max-[${TW_SCREENS.XL}px]:border-orange-500`,
```
produces this warning:

```
warn - The `min-*` and `max-*` variants are not supported with a `screens` configuration containing mixed units.
```
and styles aren't applied.

**Reply by totti-rdz** [reply]
I don't think, that even when using arbitrary values Tailwind is able to resolve dynamic values in classes.

So this is not possible 

`const color = "green"
class: "border-${color}-green"`

And I think it is not possible with `max-` either

**Reply by nemanjam** [reply]
True.

**Comment by m-shum** [comment]
How would you apply this to something like `@media (orientation: portrait) and (hover: hover)`? It works well with a single @media rule but stacking them hasn't worked for me so far. Goes without saying that I removed the spaces when writing MQs as TW classes.

**Reply by shrekuu** [reply]
The `portrait` can be utilized as follows: `<div class="h-64 bg-red-500 portrait:hover:bg-green-500"></div>`. 

In the desktop Chrome devtool, within the phone device view, it is necessary to click in order to simulate this `hover` state. 

<img width="574" alt="image" src="https://github.com/user-attachments/assets/947081bb-c0ed-4b3c-99dc-f470ca573fdd" /> 

Reference: https://tailwindcss.com/docs/hover-focus-and-other-states#viewport-orientation.



Also, you can combine them now like this, with tailwindcss 3.4.15+. `<div class="h-64 bg-red-500 portrait:hover:bg-green-500 min-[400px]:portrait:hover:bg-sky-300"></div>`


<img width="575" alt="image" src="https://github.com/user-attachments/assets/0b9cbc1d-5836-408e-b3dc-ab0b3118f413" />


reference: https://tailwindcss.com/docs/responsive-design#arbitrary-values


play it here: https://play.tailwindcss.com/dFq0O9ilLI

**Reply by m-shum** [reply]
I don't think my question was clear enough – I'm not trying to target `:hover`, I'm trying to target `@media (hover: hover)`.

**Reply by wongjn** [reply]
You could do `portrait:[@media(hover:hover)]:class-name` which would compile:
```css
@media (orientation: portrait) {
  @media (hover: hover) {
    .class-name {
```
Or you could put it all in one like `[@media(orientation:portrait)_and_(hover:hover)]:class-name` which would compile:
```css
@media (orientation: portrait) and @media (hover: hover) {
    .class-name {
```
